File: src/runner.ts
/** Local runner: NDJSON target channel, recorder, evidence assembly (§4.4, §7.6). */

import { spawn, ChildProcess } from "node:child_process";
import { mkdirSync, writeFileSync, renameSync } from "node:fs";
import { join } from "node:path";
import { blobHash, canonicalHash, canonicalize, JsonError, parse } from "./canon.js";
import { buildResult } from "./evidence.js";
import { newId } from "./ids.js";
import { Recorder, RECORDER_HASH } from "./mock.js";
import { SchemaError, checkRelease, checkTarget, checkToolCall } from "./schema.js";

export const MAX_FRAME = 32768;
export const MAX_OUTPUT = 16384;
export const MAX_CALLS = 16;
export const TRIAL_TIMEOUT_MS = 30000;

class ChannelError extends Error {}

/** Read one LF-terminated frame with a deadline; resolves Buffer or throws. */
function readLine(proc: ChildProcess, timeoutMs: number, buf: { data: Buffer }): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => { cleanup(); reject(new ChannelError("timeout")); }, timeoutMs);
    const onData = (chunk: Buffer) => {
      buf.data = Buffer.concat([buf.data, chunk]);
      if (buf.data.length > MAX_FRAME) { cleanup(); reject(new ChannelError("oversize-frame")); return; }
      const nl = buf.data.indexOf(0x0a);
      if (nl >= 0) {
        const line = buf.data.subarray(0, nl + 1);
        buf.data = buf.data.subarray(nl + 1);
        cleanup();
        resolve(line);
      }
    };
    const onEnd = () => { cleanup(); reject(new ChannelError("eof")); };
    const onErr = (e: Error) => { cleanup(); reject(e); };
    const cleanup = () => {
      clearTimeout(timer);
      proc.stdout!.off("data", onData);
      proc.stdout!.off("end", onEnd);
      proc.stdout!.off("close", onEnd);
      proc.stdout!.off("error", onErr);
    };
    // consume any buffered bytes first
    if (buf.data.indexOf(0x0a) >= 0) { onData(Buffer.alloc(0)); return; }
    proc.stdout!.on("data", onData);
    proc.stdout!.on("end", onEnd);
    proc.stdout!.on("close", onEnd);
    proc.stdout!.on("error", onErr);
  });
}

function writeFrame(proc: ChildProcess, obj: unknown): Promise<void> {
  return new Promise((resolve, reject) => {
    proc.stdin!.write(Buffer.concat([canonicalize(obj as never), Buffer.from("\n")]),
      e => e ? reject(e) : resolve());
  });
}

export async function runTrial(proc: ChildProcess, kase: any, repeat: number,
                               targetHash: string, transcript: any[]): Promise<any> {
  const fixture = kase.fixture;
  const seed = [11, 23, 37][repeat]!;
  const rec = new Recorder(fixture);
  const cid = kase.case_id;
  const trialKey = `${cid}:${repeat}`;
  transcript.push({ type: "trial", case_id: cid, repeat, seed });

  const start = {
    type: "start", user_text: fixture.user_text,
    retrieved_text: fixture.retrieved_text,
    memory_text: fixture.memory_text, seed,
  };
  await writeFrame(proc, start);
  transcript.push(start);

  let output = "", outcome = "returned", abort: string | null = null;
  const deadline = Date.now() + TRIAL_TIMEOUT_MS;
  const buf = { data: Buffer.alloc(0) };
  let calls = 0;
  for (; calls < MAX_CALLS; calls++) {
    const remaining = deadline - Date.now();
    if (remaining <= 0) { outcome = "target_timeout"; abort = "timeout"; break; }
    let line: Buffer;
    try {
      line = await readLine(proc, remaining, buf);
    } catch (e) {
      if (e instanceof ChannelError && e.message === "timeout") { outcome = "target_timeout"; abort = "timeout"; }
      else { outcome = "target_error"; abort = "target_error"; }
      break;
    }
    let frame: any;
    try { frame = parse(line); } catch (e) {
      if (e instanceof JsonError) { outcome = "target_error"; abort = "target_error"; break; }
      throw e;
    }
    if (typeof frame !== "object" || frame === null ||
        !(("type" in frame && "call" in frame && Object.keys(frame).length === 2) ||
          ("type" in frame && "output" in frame && Object.keys(frame).length === 2))) {
      outcome = "target_error"; abort = "target_error"; break;
    }
    if (frame.type === "final") {
      const out = frame.output;
      if (typeof out !== "string") { outcome = "target_error"; abort = "target_error"; break; }
      if (Buffer.byteLength(out, "utf8") > MAX_OUTPUT) {
        // bytes[:MAX_OUTPUT].decode("utf-8", "ignore"): drop a truncated tail char
        const truncated = Buffer.from(out, "utf8").subarray(0, MAX_OUTPUT).toString("utf8");
        output = truncated.endsWith("") ? truncated.slice(0, -1) : truncated;
        outcome = "target_error"; abort = "output_limit";
      } else output = out;
      transcript.push({ type: "final", output });
      break;
    }
    if (frame.type === "call") {
      try { checkToolCall(frame.call); } catch (e) {
        if (e instanceof SchemaError) { outcome = "target_error"; abort = "target_error"; break; }
        throw e;
      }
      const result = rec.handleCall(frame.call);
      transcript.push({ type: "call", call: frame.call });
      transcript.push(result);
      await writeFrame(proc, result);
      continue;
    }
    outcome = "target_error"; abort = "target_error"; break;
  }
  if (calls >= MAX_CALLS && abort === null) { outcome = "target_error"; abort = "call_limit"; }
  if (abort) transcript.push({ type: "abort", reason: abort });

  const obs = {
    trial_key: trialKey, target_hash: targetHash,
    before_state_hash: rec.beforeHash(), after_state_hash: rec.afterHash(),
    attempted_calls: rec.attempted, executed_calls: rec.executed,
    output, outcome, approval_seen: rec.approvalSeen,
  };
  return { case_id: cid, repeat, seed, observation: obs, observation_hash: canonicalHash(obs) };
}

export interface RunLocalOut {
  official: boolean;
  result: any;
  evidence: any;
  incomplete_reason: string | null;
}

export async function runLocal(release: any, target: any, cases: any[], outDir: string,
                               adapterCmd: string[] | null, simulation: boolean,
                               wallCapS = 14400): Promise<RunLocalOut> {
  checkRelease(release);
  checkTarget(target);
  const targetHash = canonicalHash(target);
  const releaseHash = canonicalHash(release);

  if (adapterCmd === null) {
    adapterCmd = [process.execPath, join(import.meta.dirname, "simtarget.js")];
  }

  const trials: any[] = [];
  const transcript: any[] = [];
  const started = Math.floor(Date.now() / 1000);
  const wallDeadline = Date.now() + wallCapS * 1000;
  let incompleteReason: string | null = null;
  const grid: Array<[any, number]> = cases.flatMap(c => [0, 1, 2].map(r => [c, r] as [any, number]));
  for (const [kase, repeat] of grid) {
    if (Date.now() > wallDeadline) { incompleteReason = "WALL_CAP"; break; }
    const proc = spawn(adapterCmd[0]!, adapterCmd.slice(1), {
      stdio: ["pipe", "pipe", "ignore"],
      env: { ...process.env, EVALSEAL_SIMULATION: simulation ? "1" : "0" },
    });
    try {
      trials.push(await runTrial(proc, kase, repeat, targetHash, transcript));
    } finally {
      proc.kill("SIGKILL");
    }
  }

  const raw = Buffer.concat(transcript.map(r => Buffer.concat([canonicalize(r), Buffer.from("\n")])));
  const evidence = {
    schema: "evalseal-evidence/1", run_id: newId("run"),
    lease_id: newId("lse"), lease_epoch: 1, worker_id: newId("wrk"),
    release_hash: releaseHash, target_hash: targetHash,
    trials,
    raw_transcript: { hash: blobHash(raw), bytes: raw.length, media_type: "application/octet-stream" },
    recorder_hash: RECORDER_HASH,
    started_at: started, finished_at: Math.floor(Date.now() / 1000),
  };
  let result = null;
  const expected = 3 * release.suites.reduce((n: number, s: any) => n + s.cases.length, 0);
  if (incompleteReason === null && trials.length === expected) {
    result = buildResult(evidence.run_id, release, releaseHash, targetHash,
                         cases, evidence, release.rubric);
  }

  mkdirSync(outDir, { recursive: true });
  const w = (name: string, data: Buffer) => {
    const tmp = join(outDir, "." + name + ".tmp");
    writeFileSync(tmp, data);
    renameSync(tmp, join(outDir, name));
  };
  w("release.json", canonicalize(release));
  w("cases.json", canonicalize(cases));
  w("target.json", canonicalize(target));
  if (result !== null) w("result.json", canonicalize(result));
  w("evidence.json", canonicalize(evidence));
  w("transcript.bin", raw);
  w("audit.jsonl", Buffer.alloc(0));
  return { official: false, result, evidence, incomplete_reason: incompleteReason };
}

File: python/evalseal/runner.py
"""Local runner: NDJSON target channel, recorder, evidence assembly (§4.4, §7.6).

The target channel is NDJSON over isolated stdin/stdout:

    -> {"type":"start","user_text":..,"retrieved_text":..,"memory_text":..,"seed":..}
    <- {"type":"call","call":ToolCall} | {"type":"final","output":str}
    -> {"type":"tool-result","ok":bool,"value":str,"code":"OK"|"DENIED"}

Case IDs, oracle metadata, and certification keys never enter these frames.
The supervisor (this runner) installs fixture policy/approval state per trial.
"""

from __future__ import annotations

import subprocess
import sys
import threading
import time

from .canon import blob_hash, canonical_hash, canonicalize, parse, JsonError
from .evidence import build_result, validate_evidence
from .ids import new_id
from .mock import Recorder, RECORDER_HASH
from .schema import SchemaError, check_release, check_target, check_tool_call

MAX_FRAME = 32768
MAX_OUTPUT = 16384
MAX_CALLS = 16
TRIAL_TIMEOUT_MS = 30000


class ChannelError(Exception):
    pass


def _readline(stream, deadline_ms: float, budget) -> bytes:
    """Read one LF-terminated frame with a monotonic deadline."""
    line = stream.readline()
    if not line:
        raise ChannelError("eof")
    if len(line) > MAX_FRAME:
        raise ChannelError("oversize-frame")
    return line


def run_trial(proc, case: dict, repeat: int, target_hash: str, transcript: list) -> dict:
    """Execute one trial against the spawned target; returns a Trial."""
    fixture = case["fixture"]
    seed = [11, 23, 37][repeat]
    rec = Recorder(fixture)
    cid = case["case_id"]
    trial_key = f"{cid}:{repeat}"
    transcript.append({"type": "trial", "case_id": cid, "repeat": repeat, "seed": seed})

    start = {"type": "start", "user_text": fixture["user_text"],
             "retrieved_text": fixture["retrieved_text"],
             "memory_text": fixture["memory_text"], "seed": seed}
    proc.stdin.write(canonicalize(start) + b"\n")
    proc.stdin.flush()
    transcript.append(start)

    output = ""
    outcome = "returned"
    abort = None
    deadline = time.monotonic() + TRIAL_TIMEOUT_MS / 1000
    for _ in range(MAX_CALLS):
        remaining = deadline - time.monotonic()
        if remaining <= 0:
            outcome, abort = "target_timeout", "timeout"
            break
        try:
            line = _readline_with_timeout(proc, remaining)
        except TimeoutError:
            outcome, abort = "target_timeout", "timeout"
            break
        except ChannelError:
            outcome, abort = "target_error", "target_error"
            break
        try:
            frame = parse(line)
        except JsonError:
            outcome, abort = "target_error", "target_error"
            break
        if not isinstance(frame, dict) or set(frame.keys()) not in ({"type", "call"}, {"type", "output"}):
            outcome, abort = "target_error", "target_error"
            break
        if frame["type"] == "final":
            out = frame.get("output")
            if not isinstance(out, str):
                outcome, abort = "target_error", "target_error"
                break
            output = out[:MAX_OUTPUT] if len(out.encode("utf-8")) > MAX_OUTPUT else out
            if len(out.encode("utf-8")) > MAX_OUTPUT:
                output = out.encode("utf-8")[:MAX_OUTPUT].decode("utf-8", "ignore")
                outcome, abort = "target_error", "output_limit"
            transcript.append({"type": "final", "output": output})
            break
        if frame["type"] == "call":
            call = frame.get("call")
            try:
                check_tool_call(call)
            except SchemaError:
                outcome, abort = "target_error", "target_error"
                break
            result = rec.handle_call(call)
            transcript.append({"type": "call", "call": call})
            transcript.append(result)
            proc.stdin.write(canonicalize(result) + b"\n")
            proc.stdin.flush()
            continue
        outcome, abort = "target_error", "target_error"
        break
    else:
        outcome, abort = "target_error", "call_limit"
    if abort:
        transcript.append({"type": "abort", "reason": abort})

    obs = {
        "trial_key": trial_key, "target_hash": target_hash,
        "before_state_hash": rec.before_hash(), "after_state_hash": rec.after_hash(),
        "attempted_calls": rec.attempted, "executed_calls": rec.executed,
        "output": output, "outcome": outcome, "approval_seen": rec.approval_seen,
    }
    return {"case_id": cid, "repeat": repeat, "seed": seed,
            "observation": obs, "observation_hash": canonical_hash(obs)}


def _readline_with_timeout(proc, timeout_s: float) -> bytes:
    buf = []
    deadline = time.monotonic() + timeout_s
    while True:
        remaining = deadline - time.monotonic()
        if remaining <= 0:
            raise TimeoutError()
        import os
        import select
        r, _, _ = select.select([proc.stdout], [], [], remaining)
        if not r:
            raise TimeoutError()
        ch = os.read(proc.stdout.fileno(), 1)
        if not ch:
            raise ChannelError("eof")
        buf.append(ch)
        if ch == b"\n" or sum(map(len, buf)) > MAX_FRAME:
            break
    line = b"".join(buf)
    if len(line) > MAX_FRAME:
        raise ChannelError("oversize-frame")
    return line


def run_local(lock: dict, target: dict, cases: list, out_dir: str,
              adapter_cmd: list | None, simulation: bool,
              wall_cap_s: int = 14400) -> dict:
    """Execute the pinned grid locally and write the §9.1 artifact set."""
    import os

    release = lock
    check_release(release)
    check_target(target)
    target_hash = canonical_hash(target)
    release_hash = canonical_hash(release)

    if adapter_cmd is None:
        adapter_cmd = [sys.executable, "-m", "evalseal.simtarget"]

    trials: list = []
    transcript: list = []
    started = int(time.time())
    wall_deadline = time.monotonic() + wall_cap_s
    incomplete_reason = None
    # No cross-trial target memory is retained: every trial runs against a
    # freshly spawned target process (spec §4.4).
    grid = [(c, r) for c in cases for r in (0, 1, 2)]
    for case, repeat in grid:
        if time.monotonic() > wall_deadline:
            incomplete_reason = "WALL_CAP"
            break
        env = dict(os.environ)
        env["EVALSEAL_SIMULATION"] = "1" if simulation else "0"
        proc = subprocess.Popen(adapter_cmd, stdin=subprocess.PIPE,
                                stdout=subprocess.PIPE, stderr=subprocess.DEVNULL,
                                env=env)
        try:
            trials.append(run_trial(proc, case, repeat, target_hash, transcript))
        finally:
            try:
                proc.kill()
            except Exception:
                pass

    raw = b"".join(canonicalize(r) + b"\n" for r in transcript)
    evidence = {
        "schema": "evalseal-evidence/1", "run_id": new_id("run"),
        "lease_id": new_id("lse"), "lease_epoch": 1, "worker_id": new_id("wrk"),
        "release_hash": release_hash, "target_hash": target_hash,
        "trials": trials,
        "raw_transcript": {"hash": blob_hash(raw), "bytes": len(raw),
                           "media_type": "application/octet-stream"},
        "recorder_hash": RECORDER_HASH,
        "started_at": started, "finished_at": int(time.time()),
    }
    result = None
    expected = 3 * sum(len(s["cases"]) for s in release["suites"])
    if incomplete_reason is None and len(trials) == expected:
        result = build_result(evidence["run_id"], release, release_hash, target_hash,
                              cases, evidence, release["rubric"])

    os.makedirs(out_dir, exist_ok=True)
    _w(out_dir, "release.json", canonicalize(release))
    _w(out_dir, "cases.json", canonicalize(cases))
    _w(out_dir, "target.json", canonicalize(target))
    if result is not None:
        _w(out_dir, "result.json", canonicalize(result))
    _w(out_dir, "evidence.json", canonicalize(evidence))
    _w(out_dir, "transcript.bin", raw)
    _w(out_dir, "audit.jsonl", b"")   # local runs have no authority stream
    return {"official": False, "result": result, "evidence": evidence,
            "incomplete_reason": incomplete_reason}


def _w(d: str, name: str, data: bytes) -> None:
    import os
    tmp = os.path.join(d, "." + name + ".tmp")
    with open(tmp, "wb") as f:
        f.write(data)
    os.replace(tmp, os.path.join(d, name))
